spring-modulith-events/spring-modulith-events-mongodb/src/main/java/org/springframework/modulith/events/mongodb/MongoDbEventPublicationRepository.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.modulith.events.mongodb;
+
+import org.springframework.data.domain.Sort;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.data.util.TypeInformation;
+import org.springframework.modulith.events.CompletableEventPublication;
+import org.springframework.modulith.events.EventPublication;
+import org.springframework.modulith.events.EventPublicationRepository;
+import org.springframework.modulith.events.PublicationTargetIdentifier;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import lombok.EqualsAndHashCode;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Repository to store {@link EventPublication}s in a MongoDB.
+ *
+ * @author Björn Kieling
+ * @author Dmitry Belyaev
+ */
+public class MongoDbEventPublicationRepository implements EventPublicationRepository {
+
+	private final MongoTemplate mongoTemplate;
+
+	public MongoDbEventPublicationRepository(MongoTemplate mongoTemplate) {
+		this.mongoTemplate = mongoTemplate;
+	}
+
+	@Override
+	public EventPublication create(EventPublication publication) {
+		mongoTemplate.save(domainToDocument(publication));
+		return publication;
+	}
+
+	@Override
+	public EventPublication update(CompletableEventPublication publication) {
+		findDocumentsByEventAndTargetIdentifier(publication.getEvent(), publication.getTargetIdentifier())
+				.stream()
+				.findFirst()
+				.ifPresent(document -> publication.getCompletionDate()
+						.ifPresent(completionDate -> {
```

Comment:
Isn't that guard here wrong? It would only set the completion date if it is already set? Wouldn't a simple `….map(MongoDbEventPublication::markCompleted).map(mongoTemplate::save)` in the line before suffice?

Reply:
The first one means that we have found a document to update. The second one means that a `completionDate` is set on the object passed as a parameter.

Changing the code the way you suggest would mean to ignore the `completionDate` value passed as a parameter.

Currently the method name is `update` what implies that the whole object will be updated which is not a case. The method name used to be `updateCompletionDate` which was a better description of what this method actually does.

It seems like removing the second condition would make this method aligned with other implementations (JPA, JDBC)

Reply:
Ahh, I see. I messed up the actual variable names. Given that the only mutable field of our persistent model is the completion date, does it make sense to early return from the method if the given `CompletableEventPublication` is not completed?

I deliberately named the original method as it too much implied that the invocation would change the completion date on the given parameter object, which is not the case at all.

Reply:
Done